### ISSUE TITLE
[Flake]: Bazel builds failed while downloading http archive.

### ISSUE THREAD

**Issue by coryan** [googleapis/google-cloud-cpp-spanner#310]
Two (otherwise unrelated) builds failed while downloading the BoringSSL archive:

https://sponge.corp.google.com/target?id=936c747b-c24b-4938-bb80-108e46618629&target=cloud-cpp%2Fspanner%2Fdocker%2Fubsan-presubmit

https://source.cloud.google.com/results/invocations/2340826a-9f4e-4277-a244-0bbacbbb84aa/targets/cloud-cpp%2Fspanner%2Fmacos%2Fbazel/log

We should consider caching the most important tarballs in GCS so the download has an alternative location. I believe the TensorFlow folks do this:

https://github.com/tensorflow/tensorflow/blob/76aa7f81ab726eb9f862c6c691ef2c76ea4e6af3/WORKSPACE#L10

**Comment by coryan** [comment]
Another one:

https://source.cloud.google.com/results/invocations/bac36efc-be8e-49c9-9501-5b8916d01520

**Comment by coryan** [comment]
And another one:

https://source.cloud.google.com/results/invocations/98f0cd7a-34df-41bd-aa77-7bdfe2b3677e

@devjgm this is the third strike on this particular sort of flakiness.

**Comment by coryan** [comment]
And another one:

https://source.cloud.google.com/results/invocations/4c714290-7d8b-4443-838b-bddd975afe84/targets

**Comment by coryan** [comment]
Here is another one of these:

https://source.cloud.google.com/results/invocations/444cb1b5-f06c-4a5e-8969-3c5ac05c8e51

**Comment by tmatsuo** [comment]
Another one:
https://source.cloud.google.com/results/invocations/47f1dde4-c6bd-4847-85d2-c5466f949efe

**Comment by coryan** [comment]
It might be possible to fix these by downloading all the dependencies (with retries) before starting the build, something like:

```sh
while i in 1 2 3; do
  if bazel fetch //google/cloud/...:all; then break; fi
  sleep 10
done
```

**Comment by devjgm** [comment]
7 days ago https://github.com/grpc/grpc/pull/21811 was merged, which I suspect will address this problem. It looks like the issue was that grpc used to be picking up boringssl without specifying a SHA256, which made it so bazel could not cache the result. The PR previously mentioned now adds a SHA and a github mirror to download the boringssl tarball. 

So I think that after the next grpc release this problem *should* go away (note: the current release is https://github.com/grpc/grpc/releases/tag/v1.27.1).

**Comment by coryan** [comment]
@devjgm if only life was that "easy".

Using a SHA256 for boringssl sounds like a great idea, but it is unlikely to help, the issue wasn't so much that the download was corrupted, rather that it was interrupted or not started and Bazel just gives up.

In #1196 we started using `bazel fetch` with a retry loop, that may have (I would say it seems to have) reduced the problems. Except that `bazel fetch` does not seem to download all the dependencies (I observe more dependencies downloaded in `bazel build`, no idea why it works like this), so I expect failures to happen, just to be more rare.

In addition, 1.27.1 was started before grpc/grpc#21811 was merged. So probably another 6 weeks before a gRPC release including that PR.

**Comment by devjgm** [comment]
> Using a SHA256 for boringssl sounds like a great idea, but it is unlikely to help, the issue wasn't so much that the download was corrupted, rather that it was interrupted or not started and Bazel just gives up.

Right. But https://github.com/grpc/grpc/pull/21811 seems to suggest that the lack of a SHA256 prevents bazel from caching the downloaded dep, and thus adding the SHA256 would allow it to cache. If bazel caches the download, then we should download boringssl less often and I'd therefore expect to see this flake less often. Well, that was my thinking anyway.

**Comment by coryan** [comment]
This is blocked until we get a gRPC version with hashes for BoringSSL, so we can remove the  🚨 label. Note, however, that do not currently cache build artifacts or downloaded dependencies between CI runs.
